File: Ghostlark/Sources/Models/ProxyNode.swift
import Foundation
import CryptoKit

enum ProxyProtocol: String, Codable, CaseIterable, Hashable {
    case vless, vmess, trojan, shadowsocks, hysteria2, tuic, wireguard

    var label: String {
        switch self {
        case .vless: return "VLESS"
        case .vmess: return "VMess"
        case .trojan: return "Trojan"
        case .shadowsocks: return "Shadowsocks"
        case .hysteria2: return "Hysteria2"
        case .tuic: return "TUIC"
        case .wireguard: return "WireGuard"
        }
    }
}

enum TLSMode: String, Codable, Hashable {
    case none, tls, reality
    var label: String {
        switch self {
        case .none: return "No TLS"
        case .tls: return "TLS"
        case .reality: return "Reality"
        }
    }
}

enum TransportKind: String, Codable, Hashable {
    case tcp, ws, grpc, http, httpupgrade, xhttp, kcp, quic

    /// sing-box does not implement Xray-only transports.
    var supportedBySingBox: Bool {
        switch self {
        case .xhttp, .kcp, .quic: return false
        default: return true
        }
    }
    var label: String { rawValue.uppercased() }
}

enum TestOutcome: Codable, Hashable {
    case untested
    case ok(Int)
    case failed(String)

    var latency: Int? { if case .ok(let ms) = self { return ms }; return nil }
    var isOK: Bool { latency != nil }
    var isFailed: Bool { if case .failed = self { return true }; return false }
}

enum SafetyTier: String, Codable, Hashable, Comparable, CaseIterable {
    case unsafe, weak, good, strong

    var order: Int {
        switch self {
        case .unsafe: return 0
        case .weak: return 1
        case .good: return 2
        case .strong: return 3
        }
    }
    static func < (lhs: SafetyTier, rhs: SafetyTier) -> Bool { lhs.order < rhs.order }

    var label: String {
        switch self {
        case .unsafe: return "Unsafe"
        case .weak: return "Weak"
        case .good: return "Good"
        case .strong: return "Strong"
        }
    }
}

struct ProxyNode: Codable, Identifiable, Hashable {
    var id: String
    var name: String
    var proto: ProxyProtocol
    var server: String
    var port: Int

    // Credentials
    var uuid: String?
    var password: String?
    var method: String?
    var flow: String?

    // TLS
    var security: TLSMode
    var sni: String?
    var alpn: [String]
    var fingerprint: String?
    var insecure: Bool
    var realityPublicKey: String?
    var realityShortId: String?

    // Transport
    var transport: TransportKind
    var path: String?
    var host: String?
    var serviceName: String?

    // Protocol specifics
    var obfsType: String?
    var obfsPassword: String?
    var congestionControl: String?
    var udpRelayMode: String?
    var alterId: Int
    var vmessSecurity: String?
    var plugin: String?
    var pluginOpts: String?

    // Provenance
    var sourceName: String
    var rawLink: String
    var firstSeen: Date
    var lastSeen: Date

    // Static assessment (filled by SafetyScorer at parse time)
    var tier: SafetyTier
    var safetyScore: Int
    var stealthScore: Int
    var notes: [String]
    var supported: Bool

    // Runtime / history
    var test: TestOutcome
    var lastTested: Date?
    var okCount: Int
    var failCount: Int
    /// Whether WARP-over-proxy (Shield) has been verified to work through this server.
    var shieldCapable: Bool?

    init(name: String, proto: ProxyProtocol, server: String, port: Int,
         uuid: String? = nil, password: String? = nil, method: String? = nil, flow: String? = nil,
         security: TLSMode = .none, sni: String? = nil, alpn: [String] = [], fingerprint: String? = nil,
         insecure: Bool = false, realityPublicKey: String? = nil, realityShortId: String? = nil,
         transport: TransportKind = .tcp, path: String? = nil, host: String? = nil, serviceName: String? = nil,
         obfsType: String? = nil, obfsPassword: String? = nil, congestionControl: String? = nil,
         udpRelayMode: String? = nil, alterId: Int = 0, vmessSecurity: String? = nil,
         plugin: String? = nil, pluginOpts: String? = nil,
         sourceName: String, rawLink: String) {
        self.name = name
        self.proto = proto
        self.server = server
        self.port = port
        self.uuid = uuid
        self.password = password
        self.method = method
        self.flow = flow
        self.security = security
        self.sni = sni
        self.alpn = alpn
        self.fingerprint = fingerprint
        self.insecure = insecure
        self.realityPublicKey = realityPublicKey
        self.realityShortId = realityShortId
        self.transport = transport
        self.path = path
        self.host = host
        self.serviceName = serviceName
        self.obfsType = obfsType
        self.obfsPassword = obfsPassword
        self.congestionControl = congestionControl
        self.udpRelayMode = udpRelayMode
        self.alterId = alterId
        self.vmessSecurity = vmessSecurity
        self.plugin = plugin
        self.pluginOpts = pluginOpts
        self.sourceName = sourceName
        self.rawLink = rawLink
        let now = Date()
        self.firstSeen = now
        self.lastSeen = now
        self.tier = .weak
        self.safetyScore = 0
        self.stealthScore = 0
        self.notes = []
        self.supported = true
        self.test = .untested
        self.lastTested = nil
        self.okCount = 0
        self.failCount = 0
        self.id = ProxyNode.makeID(proto: proto, server: server, port: port,
                                   credential: uuid ?? password ?? "", transport: transport,
                                   path: path ?? serviceName ?? "", sni: sni ?? host ?? "")
        let a = SafetyScorer.assess(self)
        self.tier = a.tier
        self.safetyScore = a.score
        self.stealthScore = a.stealth
        self.notes = a.notes
        self.supported = a.supported
    }

    static func makeID(proto: ProxyProtocol, server: String, port: Int, credential: String,
                       transport: TransportKind, path: String, sni: String) -> String {
        let key = "\(proto.rawValue)|\(server.lowercased())|\(port)|\(credential)|\(transport.rawValue)|\(path)|\(sni.lowercased())"
        let digest = SHA256.hash(data: Data(key.utf8))
        return digest.prefix(10).map { String(format: "%02x", $0) }.joined()
    }

    /// Outbound tag used inside sing-box configs.
    var tag: String { "n" + id }
    var isWARP: Bool { proto == .wireguard }

    var serverIsIP: Bool {
        if server.contains(":") { return true }
        let parts = server.split(separator: ".")
        return parts.count == 4 && parts.allSatisfy { Int($0) != nil }
    }

    /// The SNI sing-box should present, following the usual client precedence.
    var effectiveSNI: String? {
        if let s = sni, !s.isEmpty { return s }
        if let h = host, !h.isEmpty { return h }
        return serverIsIP ? nil : server
    }

    var countryFlag: String? {
        let scalars = Array(name.unicodeScalars)
        guard scalars.count >= 2 else { return nil }
        for i in 0..<(scalars.count - 1) {
            let a = scalars[i].value, b = scalars[i + 1].value
            if (0x1F1E6...0x1F1FF).contains(a) && (0x1F1E6...0x1F1FF).contains(b) {
                var s = String.UnicodeScalarView()
                s.append(scalars[i]); s.append(scalars[i + 1])
                return String(s)
            }
        }
        return nil
    }

    var countryCode: String? {
        guard let flag = countryFlag else { return nil }
        let letters = flag.unicodeScalars.map { Character(UnicodeScalar($0.value - 0x1F1E6 + 65)!) }
        return String(letters)
    }

    var reliability: Double {
        let total = okCount + failCount
        return total == 0 ? 0.5 : Double(okCount) / Double(total)
    }

    /// Composite ranking used to pick the "best" server. Higher is better.
    /// Decoy hostnames of companies that serve from their own IP ranges. A Reality handshake naming one
    /// of these while the server sits on a random VPS IP is an SNI/IP mismatch a censor can check cheaply.
    static let selfHostedDecoys = ["apple.com", "icloud.com", "microsoft.com", "live.com", "bing.com", "google.com",
        "gstatic.com", "youtube.com", "amazon.com", "cloudflare.com", "facebook.com", "instagram.com",
        "whatsapp.com", "yahoo.com", "samsung.com", "speedtest.net", "twitter.com", "x.com"]

    var decoyMismatchRisk: Bool {
        guard security == .reality, let s = (sni ?? host)?.lowercased() else { return false }
        return ProxyNode.selfHostedDecoys.contains { s == $0 || s.hasSuffix("." + $0) }
    }

    var isWebTransport: Bool { [.ws, .grpc, .http, .httpupgrade].contains(transport) }

    var cdnFronted: Bool {
        guard isWebTransport, security != .none, serverIsIP, let h = host ?? sni, !h.isEmpty else { return false }
        return h.lowercased() != server.lowercased()
    }

    /// Extra Stealth only accepts traffic that looks like ordinary HTTPS to a plausible site:
    /// Reality, or verified TLS carrying WebSocket/gRPC/HTTPUpgrade (or XTLS-Vision) on a standard HTTPS port.
    /// Shadowsocks, bare VMess, QUIC protocols, direct WireGuard and unverified certificates are excluded.
    var extraStealthEligible: Bool {
        guard supported, !isWARP else { return false }
        guard proto == .vless || proto == .trojan || proto == .vmess else { return false }
        if security == .reality { return realityPublicKey != nil }
        guard security == .tls, !insecure, SafetyScorer.commonTLSPorts.contains(port) else { return false }
        return isWebTransport || (flow?.contains("vision") ?? false)
    }

    var extraStealthScore: Int {
        var s = stealthScore
        if decoyMismatchRisk { s -= 12 }
        if cdnFronted { s += 4 }
        if !serverIsIP { s -= 4 }          // the proxy's hostname is resolved through the ISP's DNS
        if flow?.contains("vision") ?? false { s += 3 }
        return max(0, min(100, s))
    }

    func rank(stealthWeighted: Bool, shield: Bool = false, extra: Bool = false) -> Double {
        var r = Double(safetyScore) * 3 + Double(extra ? extraStealthScore : stealthScore) * (stealthWeighted ? 2.5 : 1)
        r += reliability * 120
        if shield, let s = shieldCapable { r += s ? 300 : -300 }
        switch test {
        case .ok(let ms): r += 1000 - Double(min(ms, 3000)) / 6
        case .failed: r -= 800
        case .untested: break
        }
        return r
    }

    var securityBadge: String {
        switch security {
        case .reality: return "Reality"
        case .tls: return insecure ? "TLS (unverified)" : "TLS"
        case .none: return proto == .hysteria2 || proto == .tuic || proto == .wireguard ? "Encrypted" : "Plain"
        }
    }
}

File: Android_Version/app/src/main/java/com/sepahi/ghostlark/model/ProxyNode.kt
package com.sepahi.ghostlark.model

import kotlinx.serialization.Serializable
import java.security.MessageDigest

@Serializable
enum class Proto(val label: String) {
    VLESS("VLESS"), VMESS("VMess"), TROJAN("Trojan"), SHADOWSOCKS("Shadowsocks"),
    HYSTERIA2("Hysteria2"), TUIC("TUIC"), WIREGUARD("WireGuard")
}

@Serializable
enum class TlsMode { NONE, TLS, REALITY }

@Serializable
enum class Transport(val supported: Boolean = true) {
    TCP, WS, GRPC, HTTP, HTTPUPGRADE, XHTTP(false), KCP(false), QUIC(false)
}

@Serializable
enum class Tier(val label: String) { UNSAFE("Unsafe"), WEAK("Weak"), GOOD("Good"), STRONG("Strong") }

/** One proxy server. Immutable; use copy() for updates. */
@Serializable
data class ProxyNode(
    val id: String,
    val name: String,
    val proto: Proto,
    val server: String,
    val port: Int,
    val uuid: String? = null,
    val password: String? = null,
    val method: String? = null,
    val flow: String? = null,
    val security: TlsMode = TlsMode.NONE,
    val sni: String? = null,
    val alpn: List<String> = emptyList(),
    val fingerprint: String? = null,
    val insecure: Boolean = false,
    val realityPublicKey: String? = null,
    val realityShortId: String? = null,
    val transport: Transport = Transport.TCP,
    val path: String? = null,
    val host: String? = null,
    val serviceName: String? = null,
    val obfsType: String? = null,
    val obfsPassword: String? = null,
    val congestionControl: String? = null,
    val udpRelayMode: String? = null,
    val alterId: Int = 0,
    val vmessSecurity: String? = null,
    val plugin: String? = null,
    val pluginOpts: String? = null,
    val sourceName: String,
    val rawLink: String,
    val firstSeen: Long = System.currentTimeMillis(),
    val lastSeen: Long = System.currentTimeMillis(),
    val tier: Tier = Tier.WEAK,
    val safetyScore: Int = 0,
    val stealthScore: Int = 0,
    val notes: List<String> = emptyList(),
    val supported: Boolean = true,
    val latencyMs: Int? = null,
    val lastError: String? = null,
    val lastTested: Long? = null,
    val okCount: Int = 0,
    val failCount: Int = 0,
    val shieldCapable: Boolean? = null,
) {
    val tag: String get() = "n$id"
    val isWarp: Boolean get() = proto == Proto.WIREGUARD
    val testedOk: Boolean get() = latencyMs != null

    val serverIsIp: Boolean
        get() = server.contains(":") || server.split(".").let { it.size == 4 && it.all { p -> p.toIntOrNull() != null } }

    val effectiveSni: String?
        get() = sni?.takeIf { it.isNotEmpty() } ?: host?.takeIf { it.isNotEmpty() } ?: if (serverIsIp) null else server

    val countryFlag: String?
        get() {
            val cps = name.codePoints().toArray()
            for (i in 0 until cps.size - 1) {
                if (cps[i] in 0x1F1E6..0x1F1FF && cps[i + 1] in 0x1F1E6..0x1F1FF) {
                    return String(intArrayOf(cps[i], cps[i + 1]), 0, 2)
                }
            }
            return null
        }

    val countryCode: String?
        get() = countryFlag?.codePoints()?.toArray()?.map { (it - 0x1F1E6 + 'A'.code).toChar() }?.joinToString("")

    val reliability: Double
        get() = if (okCount + failCount == 0) 0.5 else okCount.toDouble() / (okCount + failCount)

    /** A Reality decoy naming a site that serves from its own IP ranges is an SNI/IP mismatch a censor can check. */
    val decoyMismatchRisk: Boolean
        get() {
            if (security != TlsMode.REALITY) return false
            val s = (sni ?: host)?.lowercase() ?: return false
            return selfHostedDecoys.any { s == it || s.endsWith(".$it") }
        }

    val isWebTransport: Boolean get() = transport in setOf(Transport.WS, Transport.GRPC, Transport.HTTP, Transport.HTTPUPGRADE)

    val cdnFronted: Boolean
        get() {
            val h = host ?: sni
            return isWebTransport && security != TlsMode.NONE && serverIsIp && !h.isNullOrEmpty() && !h.equals(server, true)
        }

    /** Extra Stealth: Reality, or verified TLS carrying WebSocket/gRPC/HTTPUpgrade (or Vision) on an HTTPS port. */
    val extraStealthEligible: Boolean
        get() {
            if (!supported || isWarp) return false
            if (proto != Proto.VLESS && proto != Proto.TROJAN && proto != Proto.VMESS) return false
            if (security == TlsMode.REALITY) return realityPublicKey != null
            if (security != TlsMode.TLS || insecure || port !in tlsPorts) return false
            return isWebTransport || flow?.contains("vision") == true
        }

    val extraStealthScore: Int
        get() {
            var s = stealthScore
            if (decoyMismatchRisk) s -= 12
            if (cdnFronted) s += 4
            if (!serverIsIp) s -= 4
            if (flow?.contains("vision") == true) s += 3
            return s.coerceIn(0, 100)
        }

    fun rank(stealthWeighted: Boolean, shield: Boolean = false, extra: Boolean = false): Double {
        var r = safetyScore * 3.0 + (if (extra) extraStealthScore else stealthScore) * (if (stealthWeighted) 2.5 else 1.0)
        r += reliability * 120
        if (shield && shieldCapable != null) r += if (shieldCapable) 300 else -300
        val ms = latencyMs
        if (ms != null) r += 1000 - minOf(ms, 3000) / 6.0
        else if (lastError != null) r -= 800
        return r
    }

    val securityBadge: String
        get() = when (security) {
            TlsMode.REALITY -> "Reality"
            TlsMode.TLS -> if (insecure) "TLS (unverified)" else "TLS"
            TlsMode.NONE -> if (proto == Proto.HYSTERIA2 || proto == Proto.TUIC || proto == Proto.WIREGUARD) "Encrypted" else "Plain"
        }

    /** Returns a copy with the static safety/stealth assessment filled in and a stable id. */
    fun assessed(): ProxyNode {
        val a = SafetyScorer.assess(this)
        return copy(
            id = makeId(proto, server, port, uuid ?: password ?: "", transport, path ?: serviceName ?: "", sni ?: host ?: ""),
            tier = a.tier, safetyScore = a.score, stealthScore = a.stealth, notes = a.notes, supported = a.supported,
        )
    }

    companion object {
        const val WARP_ID = "warp0000000000000000"
        val tlsPorts = setOf(443, 2053, 2083, 2087, 2096, 8443)
        val selfHostedDecoys = listOf("apple.com", "icloud.com", "microsoft.com", "live.com", "bing.com", "google.com",
            "gstatic.com", "youtube.com", "amazon.com", "cloudflare.com", "facebook.com", "instagram.com",
            "whatsapp.com", "yahoo.com", "samsung.com", "speedtest.net", "twitter.com", "x.com")

        fun makeId(proto: Proto, server: String, port: Int, credential: String, transport: Transport, path: String, sni: String): String {
            val key = "${proto.name.lowercase()}|${server.lowercase()}|$port|$credential|${transport.name.lowercase()}|$path|${sni.lowercase()}"
            val digest = MessageDigest.getInstance("SHA-256").digest(key.toByteArray())
            return digest.take(10).joinToString("") { "%02x".format(it) }
        }

        fun warpNode(): ProxyNode = ProxyNode(
            id = WARP_ID, name = "🌐 Cloudflare WARP (free, direct)", proto = Proto.WIREGUARD,
            server = "engage.cloudflareclient.com", port = 2408, sourceName = "Cloudflare WARP", rawLink = "warp://cloudflare",
        ).let { n -> val a = SafetyScorer.assess(n); n.copy(tier = a.tier, safetyScore = a.score, stealthScore = a.stealth, notes = a.notes) }
    }
}
